File: feature_tests/dart/lib/src/One.dart
import 'Two.dart';
import 'dart:ffi' as ffi;
import 'diplomat_runtime.dart';

typedef OneFfi = ffi.Pointer<ffi.Opaque>;

class One implements ffi.Finalizable {
  final OneFfi _underlying;

  factory One.transitivity(One hold, One nohold) {
    final result = _transitivityFfi(OneAsFfi(hold), OneAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _transitivityFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi)>>(
              'One_transitivity')
          .asFunction<OneFfi Function(OneFfi, OneFfi)>();

  factory One.cycle(Two hold, One nohold) {
    final result = _cycleFfi(TwoAsFfi(hold), OneAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _cycleFfi =
      capi<ffi.NativeFunction<OneFfi Function(TwoFfi, OneFfi)>>('One_cycle')
          .asFunction<OneFfi Function(TwoFfi, OneFfi)>();

  factory One.manyDependents(One a, One b, Two c, Two d, Two nohold) {
    final result = _manyDependentsFfi(
        OneAsFfi(a), OneAsFfi(b), TwoAsFfi(c), TwoAsFfi(d), TwoAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _manyDependentsFfi = capi<
          ffi.NativeFunction<
              OneFfi Function(OneFfi, OneFfi, TwoFfi, TwoFfi,
                  TwoFfi)>>('One_many_dependents')
      .asFunction<OneFfi Function(OneFfi, OneFfi, TwoFfi, TwoFfi, TwoFfi)>();

  factory One.returnOutlivesParam(Two hold, One nohold) {
    final result = _returnOutlivesParamFfi(TwoAsFfi(hold), OneAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _returnOutlivesParamFfi =
      capi<ffi.NativeFunction<OneFfi Function(TwoFfi, OneFfi)>>(
              'One_return_outlives_param')
          .asFunction<OneFfi Function(TwoFfi, OneFfi)>();

  factory One.diamondTop(One top, One left, One right, One bottom) {
    final result = _diamondTopFfi(
        OneAsFfi(top), OneAsFfi(left), OneAsFfi(right), OneAsFfi(bottom));
    return OneFromFfi(result);
  }
  static late final _diamondTopFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>>(
              'One_diamond_top')
          .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>();

  factory One.diamondLeft(One top, One left, One right, One bottom) {
    final result = _diamondLeftFfi(
        OneAsFfi(top), OneAsFfi(left), OneAsFfi(right), OneAsFfi(bottom));
    return OneFromFfi(result);
  }
  static late final _diamondLeftFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>>(
              'One_diamond_left')
          .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>();

  factory One.diamondRight(One top, One left, One right, One bottom) {
    final result = _diamondRightFfi(
        OneAsFfi(top), OneAsFfi(left), OneAsFfi(right), OneAsFfi(bottom));
    return OneFromFfi(result);
  }
  static late final _diamondRightFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>>(
              'One_diamond_right')
          .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>();

  factory One.diamondBottom(One top, One left, One right, One bottom) {
    final result = _diamondBottomFfi(
        OneAsFfi(top), OneAsFfi(left), OneAsFfi(right), OneAsFfi(bottom));
    return OneFromFfi(result);
  }
  static late final _diamondBottomFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>>(
              'One_diamond_bottom')
          .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>();

  factory One.diamondAndNestedTypes(One a, One b, One c, One d, One nohold) {
    final result = _diamondAndNestedTypesFfi(
        OneAsFfi(a), OneAsFfi(b), OneAsFfi(c), OneAsFfi(d), OneAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _diamondAndNestedTypesFfi = capi<
          ffi.NativeFunction<
              OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi,
                  OneFfi)>>('One_diamond_and_nested_types')
      .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi, OneFfi)>();

  factory One.implicitBounds(One explicitHold, One implicitHold, One nohold) {
    final result = _implicitBoundsFfi(
        OneAsFfi(explicitHold), OneAsFfi(implicitHold), OneAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _implicitBoundsFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi, OneFfi)>>(
              'One_implicit_bounds')
          .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi)>();

  factory One.implicitBoundsDeep(
      One explicit, One implicit1, One implicit2, One nohold) {
    final result = _implicitBoundsDeepFfi(OneAsFfi(explicit),
        OneAsFfi(implicit1), OneAsFfi(implicit2), OneAsFfi(nohold));
    return OneFromFfi(result);
  }
  static late final _implicitBoundsDeepFfi =
      capi<ffi.NativeFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>>(
              'One_implicit_bounds_deep')
          .asFunction<OneFfi Function(OneFfi, OneFfi, OneFfi, OneFfi)>();

  One._(this._underlying) {
    _finalizer.attach(this, this._underlying.cast());
  }

  static late final _finalizer = ffi.NativeFinalizer(
      capi<ffi.NativeFunction<ffi.Void Function(ffi.Pointer<ffi.Void>)>>(
          'One_destroy'));
}

// These are not methods because we want to keep them package-private, and methods are either private or public
One OneFromFfi(OneFfi underlying) => One._(underlying);
OneFfi OneAsFfi(One t) => t._underlying;

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/One.kt
package dev.gigapixel.somelib;
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer


internal interface OneLib: Library {
    fun One_destroy(handle: Long)
    fun One_transitivity(hold: Long, nohold: Long): Long
    fun One_cycle(hold: Long, nohold: Long): Long
    fun One_many_dependents(a: Long, b: Long, c: Long, d: Long, nohold: Long): Long
    fun One_return_outlives_param(hold: Long, nohold: Long): Long
    fun One_diamond_top(top: Long, left: Long, right: Long, bottom: Long): Long
    fun One_diamond_left(top: Long, left: Long, right: Long, bottom: Long): Long
    fun One_diamond_right(top: Long, left: Long, right: Long, bottom: Long): Long
    fun One_diamond_bottom(top: Long, left: Long, right: Long, bottom: Long): Long
    fun One_diamond_and_nested_types(a: Long, b: Long, c: Long, d: Long, nohold: Long): Long
    fun One_implicit_bounds(explicitHold: Long, implicitHold: Long, nohold: Long): Long
    fun One_implicit_bounds_deep(explicit: Long, implicit1: Long, implicit2: Long, nohold: Long): Long
}

class One internal constructor (
    internal val handle: Long,

    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal val aEdges: List<Any>,
    ) {

    internal class OneCleaner(val handle: Long, val lib: OneLib) : Runnable {
        override fun run() {
            lib.One_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<OneLib> = OneLib::class.java
        internal val lib: OneLib = Native.load("somelib", libClass)
        fun transitivity(hold: One, nohold: One): One {
            
            val returnVal = lib.One_transitivity(hold.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val aEdges: List<Any> = listOf(hold)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun cycle(hold: Two, nohold: One): One {
            
            val returnVal = lib.One_cycle(hold.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val aEdges: List<Any> = listOf(hold)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun manyDependents(a: One, b: One, c: Two, d: Two, nohold: Two): One {
            
            val returnVal = lib.One_many_dependents(a.handle, b.handle, c.handle, d.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val aEdges: List<Any> = listOf(a, b, c, d)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun returnOutlivesParam(hold: Two, nohold: One): One {
            
            val returnVal = lib.One_return_outlives_param(hold.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val longEdges: List<Any> = listOf(hold)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, longEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun diamondTop(top: One, left: One, right: One, bottom: One): One {
            
            val returnVal = lib.One_diamond_top(top.handle, left.handle, right.handle, bottom.handle);
        
            val selfEdges: List<Any> = listOf()
            val topEdges: List<Any> = listOf(top, left, right, bottom)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, topEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun diamondLeft(top: One, left: One, right: One, bottom: One): One {
            
            val returnVal = lib.One_diamond_left(top.handle, left.handle, right.handle, bottom.handle);
        
            val selfEdges: List<Any> = listOf()
            val leftEdges: List<Any> = listOf(left, bottom)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, leftEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun diamondRight(top: One, left: One, right: One, bottom: One): One {
            
            val returnVal = lib.One_diamond_right(top.handle, left.handle, right.handle, bottom.handle);
        
            val selfEdges: List<Any> = listOf()
            val rightEdges: List<Any> = listOf(right, bottom)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, rightEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun diamondBottom(top: One, left: One, right: One, bottom: One): One {
            
            val returnVal = lib.One_diamond_bottom(top.handle, left.handle, right.handle, bottom.handle);
        
            val selfEdges: List<Any> = listOf()
            val bottomEdges: List<Any> = listOf(bottom)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, bottomEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun diamondAndNestedTypes(a: One, b: One, c: One, d: One, nohold: One): One {
            
            val returnVal = lib.One_diamond_and_nested_types(a.handle, b.handle, c.handle, d.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val aEdges: List<Any> = listOf(a, b, c, d)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun implicitBounds(explicitHold: One, implicitHold: One, nohold: One): One {
            
            val returnVal = lib.One_implicit_bounds(explicitHold.handle, implicitHold.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val aEdges: List<Any> = listOf(explicitHold, implicitHold)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
        fun implicitBoundsDeep(explicit: One, implicit1: One, implicit2: One, nohold: One): One {
            
            val returnVal = lib.One_implicit_bounds_deep(explicit.handle, implicit1.handle, implicit2.handle, nohold.handle);
        
            val selfEdges: List<Any> = listOf()
            val aEdges: List<Any> = listOf(explicit, implicit1, implicit2)
            val handle = returnVal
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            
            return returnOpaque
        
        }
    }

}
